tools/autograd/load_derivatives.py

```diff
@@ -1,115 +1,115 @@
+    # that are semantically related.
+    functions_by_signature: Dict[FunctionSchema, List[NativeFunction]] = defaultdict(list)
+    functions_by_schema: Dict[str, NativeFunction] = dict()
+    for function in functions:
+        functions_by_signature[function.func.signature()].append(function)
+        assert str(function.func) not in functions_by_schema
+        functions_by_schema[str(function.func)] = function
+
+    infos = [
+        create_differentiability_info(defn, functions_by_signature, functions_by_schema)
         for defn in definitions]
 
-    autograd_functions = [d['autograd_fn'] for d in differentiability_infos if d['autograd_fn'] is not None]
-    ensure_unique_names(autograd_functions)
-    match_declarations_with_differentiability_info(declarations, differentiability_infos)
-
-    return autograd_functions
-
-
-def create_differentiability_info(signature, non_differentiable_arg_names,
-                                  output_differentiability,
-                                  autograd_fn):
-    return {
-        'signature': signature,
-        'non_differentiable_arg_names': non_differentiable_arg_names,
-        'output_differentiability': output_differentiability,
-        'autograd_fn': autograd_fn,
-    }
-
-
-# How do you feel about pasting declaration inside autograd function...
-def create_autograd_function(name, derivatives, args_with_derivatives,
-                             declaration):
-    op = to_camel_case(name) + 'Backward'
-    op = op.replace('ForwardBackward', 'Backward')
-    return {
-        'name': name,
-        'op': op,
-        'declaration': declaration,
-        'args_with_derivatives': args_with_derivatives,
-        'derivatives': derivatives,
-        'saved_inputs': all_saved_variables(derivatives, 'saved_inputs'),
-        'saved_outputs': all_saved_variables(derivatives, 'saved_outputs'),
-    }
-
-
-def create_derivative(arguments, returns, name, formula, var_names):
-    def transform_return(r):
-        # In-place functions take in and return self. Call the modified version
-        # "output" so that it can be referred to in derivative definitions.
-        if r['name'] == 'self':
-            r = copy.deepcopy(r)
-            r['name'] = 'result'
-        return r
-
-    returns = [transform_return(r) for r in returns]
-    formula, saved_inputs = saved_variables(formula, arguments, var_names)
-    formula, saved_outputs = saved_variables(formula, returns, var_names)
+    # To keep it byte-for-byte compatible with the old codegen, we assign op names as a separate
+    # step. We only assign op names to those with differentiable args, and only append suffix to
+    # duplicated op names. This can be simplified if the first of the duplicates can be named
+    # 'XyzBackward' instead of 'XyzBackward0' or unconditionally append '0' to singletons.
+    op_names = create_op_names(infos)
+    return [
+        DifferentiabilityInfo(
```

Comment:
This is just to add the op_name to each DifferentiabilityInfo but because they're unmutable, we need to recreate tall of them right?

Reply:
correct